File: Data Structures & Algorithms/binary-tree-diameter/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0
        if not root:
            return res
        def height(root):
            if not root:
                return 0
            leftHeight=height(root.left)
            rightHeight=height(root.right)
            total = leftHeight + rightHeight
            nonlocal res
            res = max(total,res)
            return 1+max(height(root.left),height(root.right))
        
        height(root)
        return res
        
        

File: Data Structures & Algorithms/max-water-container/submission-5.java
class Solution {
    public int maxArea(int[] heights) {
        int max = 0;
        int left = heights[0];
        int right = heights[heights.length-1];
        int l = 0;
        int r = heights.length - 1;
        while (l<r){
            if (heights[l]<heights[r]){
                int width = r-l;
                int area = heights[l]*width;
                max = Math.max(area,max);
                l++;
            }
            else{int width = r-l;
                int area = heights[r]*width;
                max = Math.max(area,max);
                r--;

            }
        }
        return max;
        
    }
}
